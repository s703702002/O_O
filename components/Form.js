import React, { Component } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

const FormControl = ({
  id = null,
  type = 'text',
  label = '標題文字',
  placeholder = '',
  onChange = null,
  onBlur = null,
  isInvalid = '',
  defaultValue = null,
}) => (
  <React.Fragment>
    <label htmlFor={id}>{label}</label>
    <input
      className={cx('form-control', {
        'is-invalid': isInvalid.length > 0,
      })}
      type={type}
      id={id}
      placeholder={placeholder}
      onChange={onChange}
      onBlur={onBlur}
      defaultValue={defaultValue}
    />
    {
      (isInvalid.length > 0) ?
        <div className="invalid-feedback">
          {isInvalid}
        </div> :
        null
    }
  </React.Fragment>
);

const FormRow = ({ children }) => (
  <div className="form-row">
    {children}
  </div>
);

const FormGroup = ({
  children,
  md,
  xl,
}) => (
  <div
    className={cx('form-grop', {
      [`col-md-${md}`]: md > 0,
      [`col-md-${xl}`]: xl > 0,
    })}
  >
    {children}
  </div>
);

class Form extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.node,
    ]).isRequired,
    className: PropTypes.string,
  };
  static defaultProps = {
    className: null,
  };
  static Row = FormRow;
  static Group = FormGroup;
  static Control = FormControl;
  render() {
    const {
      children,
      className,
    } = this.props;
    return (
      <form className={className}>
        {children}
      </form>
    );
  }
}


export default Form;
