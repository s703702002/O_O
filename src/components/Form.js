import React, { Component } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import cityData from '../static/taiwan.json';
import { clone } from '../utilis';

function formCheck(id, value) {
  switch (id) {
    case 'name':
    case 'address':
      return (/[a-zA-Z\u4e00-\u9fa5]/g.test(value)) ?
        '' :
        '至少輸入一個中文或英文字';
    case 'cellphone':
      return (/^09\d{8}/g.test(value)) ?
        '' :
        '請輸入正確的手機格式';
    default:
      return '';
  }
}

const FormControl = ({
  id = null,
  type = 'text',
  label = '標題文字',
  placeholder = '',
  onChange = null,
  onBlur = null,
  isInvalid = '',
  defaultValue = null,
  pattern = null
}) => (
  <React.Fragment>
    <label htmlFor={id}>{label}</label>
    <input
      className={cx('form-control', {
        'is-invalid': isInvalid.length > 0,
      })}
      type={type}
      id={id}
      placeholder={placeholder}
      onChange={onChange}
      onBlur={onBlur}
      defaultValue={defaultValue}
      pattern={pattern}
    />
    {
      (isInvalid.length > 0) ?
        <div className="invalid-feedback">
          {isInvalid}
        </div> :
        null
    }
  </React.Fragment>
);

const FormRow = ({ children }) => (
  <div className="form-row">
    {children}
  </div>
);

const FormGroup = ({
  children,
  md,
  xl,
}) => (
  <div
    className={cx('form-grop', {
      [`col-md-${md}`]: md > 0,
      [`col-md-${xl}`]: xl > 0,
    })}
  >
    {children}
  </div>
);

class Form extends Component {
  static propTypes = {
    className: PropTypes.string,
  };
  static defaultProps = {
    className: null,
  };
  static Row = FormRow;
  static Group = FormGroup;
  static Control = FormControl;
  state = {
    name: {
      value: '',
      error: '',
    },
    cellphone: {
      value: '',
      error: '',
    },
    address: {
      value: '',
      error: '',
    },
    remark: {
      value: '',
      error: '',
    },
    cityValue: 0,
    regionValue: 0,
    isValid: false,
  };
  getValid() {
    return this.valueCheckAll();
  }
  valueCheckAll = () => {
    const copy = clone(this.state);
    const keys = Object.keys(copy);
    let isValid = true;
    keys.forEach((id) => {
      const { value } = copy[id];

      // 代表是不需驗證的欄位
      if (typeof value === 'undefined') return;

      copy[id] = {
        value,
        error: formCheck(id, value),
      };

      if (copy[id].error.length > 0) isValid = false;
    });
    this.setState(() => ({
      ...copy,
      isValid,
    }));
    return isValid;
  }
  formChange = (e) => {
    const {
      id,
      value,
    } = e.target;

    this.setState(prevState => ({
      ...prevState,
      [id]: {
        value,
        error: '',
      },
    }));
  }
  valueCheck = (e) => {
    const {
      id,
      value,
    } = e.target;

    const error = formCheck(id, value);

    if (error.length) {
      this.setState(prevState => ({
        ...prevState,
        [id]: {
          value,
          error,
        },
        isValid: false,
      }));
    }
  }
  selectChange = (e) => {
    const {
      id,
      value,
    } = e.target;
    this.setState(prevState => ({
      ...prevState,
      regionValue: 0,
      [id]: value,
    }));
  }
  render() {
    const {
      className,
    } = this.props;

    const {
      name,
      cellphone,
      address,
      remark,
      cityValue,
      regionValue,
    } = this.state;

    const region = cityData.region[cityValue]; // 鄉鎮地區

    return (
      <form className={className}>
        <Form.Row>
          <Form.Group md={6}>
            <Form.Control
              label="姓名"
              id="name"
              placeholder="陳阿三"
              value={name.value}
              isInvalid={name.error}
              onChange={this.formChange}
              onBlur={this.valueCheck}
            />
          </Form.Group>
          <Form.Group md={6}>
            <Form.Control
              id="cellphone"
              label="手機"
              placeholder="0912345678"
              type="tel"
              value={cellphone.value}
              isInvalid={cellphone.error}
              onChange={this.formChange}
              onBlur={this.valueCheck}
              pattern="09[0-9]{8}"
            />
          </Form.Group>
        </Form.Row>
        <Form.Row>
          <Form.Group md={6}>
            <label htmlFor="cityValue">縣市</label>
            <select id="cityValue" className="form-control" defaultValue={cityValue} onChange={this.selectChange}>
              {
                cityData.city.map((c, i) => <option value={i} key={c}>{c}</option>)
              }
            </select>
          </Form.Group>
          <Form.Group md={4}>
            <label htmlFor="regionValue">鄉鎮地區</label>
            <select id="regionValue" className="form-control" value={regionValue} onChange={this.selectChange}>
              {
                region.map((r, i) => <option value={i} key={r}>{r}</option>)
              }
            </select>
          </Form.Group>
          <Form.Group md={2}>
            <Form.Control
              id="inputZip"
              label="郵遞區號"
              defaultValue="114"
            />
          </Form.Group>
        </Form.Row>
        <Form.Group>
          <Form.Control
            id="address"
            label="地址"
            value={address.value}
            isInvalid={address.error}
            onChange={this.formChange}
            onBlur={this.valueCheck}
          />
        </Form.Group>
        <Form.Group>
          <label htmlFor="remark">備註</label>
          <textarea
            className="form-control"
            id="remark"
            rows="3"
            value={remark.value}
            onChange={this.formChange}
          />
        </Form.Group>
      </form>
    );
  }
}


export default Form;
