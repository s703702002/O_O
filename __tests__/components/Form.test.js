import React, {useReducer, useCallback} from 'react';
import {
  render,
  fireEvent,
  cleanup,
} from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks'
import Form, { useFormState } from '../../src/components/Form'

afterEach(cleanup);

test('Form render correct', () => {
  const { result } = renderHook(() => useFormState())
  const { getByLabelText, getByDisplayValue, container, asFragment } = render(
    <Form
      state={result.current.state}
      formChange={result.current.formChange}
      selectChange={result.current.selectChange}
    />
  )

  act(() => {
    expect(result.current.validateForm()).toBeFalsy();
  })

  const nameInput = getByLabelText('姓名')
  act(() => {
    fireEvent.change(nameInput, { target: { value: 'acccdbc' } });
  })
  expect(result.current.state.name.value).toBe('acccdbc')

  const cellphoneInput = getByLabelText('手機');
  act(() => {
    fireEvent.change(cellphoneInput, { target: { value: '0912345678' } });
  })
  expect(result.current.state.cellphone.value).toBe('0912345678');

  const addressInput = getByLabelText('地址');
  act(() => {
    fireEvent.change(addressInput, { target: { value: 'anywhere' } });
  })
  expect(result.current.state.address.value).toBe('anywhere');

  act(() => {
    expect(result.current.validateForm()).toBeTruthy();
  })

  expect(asFragment()).toMatchSnapshot()
});
