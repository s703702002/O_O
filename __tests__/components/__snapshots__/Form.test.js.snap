// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render correct 1`] = `
<DocumentFragment>
  <form
    class="customer_info_form"
    id=""
  >
    <div
      class="form-row"
    >
      <div
        class="form-grop col-md-6"
      >
        <label
          for="name"
        >
          姓名
        </label>
        <input
          class="form-control"
          id="name"
          placeholder="陳阿三"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-grop col-md-6"
      >
        <label
          for="cellphone"
        >
          手機
        </label>
        <input
          class="form-control"
          id="cellphone"
          pattern="09[0-9]{8}"
          placeholder="0912345678"
          type="tel"
          value=""
        />
      </div>
    </div>
    <div
      class="form-row"
    >
      <div
        class="form-grop col-md-6"
      >
        <label
          for="cityValue"
        >
          縣市
        </label>
        <select
          class="form-control"
          id="cityValue"
        >
          <option
            value="0"
          >
            臺北市
          </option>
          <option
            value="1"
          >
            基隆市
          </option>
          <option
            value="2"
          >
            新北市
          </option>
          <option
            value="3"
          >
            宜蘭縣
          </option>
          <option
            value="4"
          >
            桃園市
          </option>
          <option
            value="5"
          >
            新竹市
          </option>
          <option
            value="6"
          >
            新竹縣
          </option>
          <option
            value="7"
          >
            苗栗縣
          </option>
          <option
            value="8"
          >
            臺中市
          </option>
          <option
            value="9"
          >
            彰化縣
          </option>
          <option
            value="10"
          >
            南投縣
          </option>
          <option
            value="11"
          >
            嘉義市
          </option>
          <option
            value="12"
          >
            嘉義縣
          </option>
          <option
            value="13"
          >
            雲林縣
          </option>
          <option
            value="14"
          >
            臺南市
          </option>
          <option
            value="15"
          >
            高雄市
          </option>
          <option
            value="16"
          >
            澎湖縣
          </option>
          <option
            value="17"
          >
            金門縣
          </option>
          <option
            value="18"
          >
            屏東縣
          </option>
          <option
            value="19"
          >
            臺東縣
          </option>
          <option
            value="20"
          >
            花蓮縣
          </option>
          <option
            value="21"
          >
            連江縣
          </option>
        </select>
      </div>
      <div
        class="form-grop col-md-4"
      >
        <label
          for="regionValue"
        >
          鄉鎮地區
        </label>
        <select
          class="form-control"
          id="regionValue"
        >
          <option
            value="0"
          >
            中正區
          </option>
          <option
            value="1"
          >
            大同區
          </option>
          <option
            value="2"
          >
            中山區
          </option>
          <option
            value="3"
          >
            松山區
          </option>
          <option
            value="4"
          >
            大安區
          </option>
          <option
            value="5"
          >
            萬華區
          </option>
          <option
            value="6"
          >
            信義區
          </option>
          <option
            value="7"
          >
            士林區
          </option>
          <option
            value="8"
          >
            北投區
          </option>
          <option
            value="9"
          >
            內湖區
          </option>
          <option
            value="10"
          >
            南港區
          </option>
          <option
            value="11"
          >
            文山區
          </option>
        </select>
      </div>
      <div
        class="form-grop col-md-2"
      >
        <label
          for="inputZip"
        >
          郵遞區號
        </label>
        <input
          class="form-control"
          id="inputZip"
          placeholder=""
          type="text"
          value="114"
        />
      </div>
    </div>
    <div
      class="form-row"
    >
      <div
        class="form-grop col-md-12"
      >
        <label
          for="address"
        >
          地址
        </label>
        <input
          class="form-control"
          id="address"
          placeholder=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="form-row"
    >
      <div
        class="form-grop col-md-12"
      >
        <label
          for="remark"
        >
          備註
        </label>
        <textarea
          class="form-control"
          id="remark"
          rows="3"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;
